File: C/LinkList/LinkList.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

//学生信息结构
typedef struct _student
{
	int id;
	char name[16];
}STUDENT, *LPSTUDENT;

//链表结构
typedef struct _link_list
{
	STUDENT s;           //链表中的节点的数据项，这里用一个结构体表示
	_link_list* next;    //指向下一个节点
}LINKLIST,*LPLINKLIST;

LINKLIST head = { 0 };   //创建一个全局的链表头节点(可以用一个函数创建，返回一个局部头节点)
int g_iCount = 0;        //记录当前链表中元素的个数

					
//判断表是否为空
bool IsEmpty()
{
	return g_iCount ? false : true;
}

//获取元素个数
int GetElementCount()
{
	return g_iCount;
}

//添加数据（头添加）
void AddFromHead(LPLINKLIST l)
{
	l->next = head.next;    //例如 head->l1,现在添加了个l,则l->l1,head->l,就是改变指向
	head.next = l;
	++g_iCount;
}

//添加数据（尾添加）
void AddFromTail(LPLINKLIST l)
{
	LPLINKLIST cur = &head;   
	while (cur->next)         //千万不能写成while(cur=cur->next);，因为当遍历完之后再cur->next(空指针)会出错
		cur = cur->next;      //遍历到链表末尾
	cur->next = l;            //把l挂在链表的末尾
	l->next = NULL;           //最后让l指向NULL
	++g_iCount;
}

//插入数据（指定位置）
void Insert(int local, LPLINKLIST l)
{
	if (local < g_iCount && local >= 0)   //当要插入的位置合理的情况下才能插入
	{
		LPLINKLIST cur = &head;
		for (int i = 0;i < local;i++)    //遍历到指定位置
			cur = cur->next;
		l->next = cur->next;              //在该位置插入
		cur->next = l;
		++g_iCount;
	}
}

//根据id返回元素位置
int GetLocalFromId(int id)
{
	LPLINKLIST cur = head.next;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (cur->s.id == id)
			return i;
		cur = cur->next;
	}
	return -1;   //不存在，返回-1
}

//根据name返回元素位置
int GetLocalFromName(char* name)
{
	LPLINKLIST cur = head.next;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (strcmp(cur->s.name,name) == 0)
			return i;
		cur = cur->next;
	}
	return -1;   //不存在，返回-1
}

//删除数据 （指定位置）
void DeleteFromLocal(int local)
{
	if (local < g_iCount && local >= 0)   //当要删除的位置合理的情况下才能删除
	{
		LPLINKLIST cur = &head;
		for (int i = 0;i < local;i++)    //遍历到指定位置
			cur = cur->next;
		LPLINKLIST temp;
		temp = cur->next;
		cur->next = temp->next;           //删除该位置下的元素
		--g_iCount;
		free(temp);                       //释放掉该节点
	}
}

//删除数据 （指定id）
void DeleteFromId(int id)
{
	LPLINKLIST cur = head.next;
	LPLINKLIST pre = &head;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (cur->s.id == id)          //找到该节点
		{
			LPLINKLIST temp;
			temp = cur;
			pre->next = temp->next;   //修改指向，就是删除了   
			free(temp);               //最后要释放节点
			--g_iCount;               
			break;                    //记得退出，否则访问空指针会出错
		}
		cur = cur->next;
		pre = pre->next;
	}
}

//删除数据 （指定name）
void DeleteFromName(char* name)
{
	LPLINKLIST cur = head.next;
	LPLINKLIST pre = &head;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (strcmp(cur->s.name, name) == 0)          //找到该节点
		{
			LPLINKLIST temp;
			temp = cur;
			pre->next = temp->next;   //修改指向，就是删除了   
			free(temp);               //最后要释放节点
			--g_iCount;
			break;
		}
		cur = cur->next;
		pre = pre->next;
	}
}

//查找数据 （指定id，返回name）
char* SearchFromId(int id)
{
	LPLINKLIST cur = head.next;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (cur->s.id == id)          //找到该节点
			return cur->s.name;
		cur = cur->next;
	}
	return NULL;
}

//查找数据 （指定name，返回id）
int SearchFromName(char* name)
{
	LPLINKLIST cur = head.next;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (strcmp(cur->s.name,name) == 0)          //找到该节点
			return cur->s.id;
		cur = cur->next;
	}
	return -1;
}

//根据local查找指定节点，返回该节点
LPLINKLIST GetNode(int local)
{
	if (local < g_iCount && local >= 0)
	{
		LPLINKLIST cur = head.next;
		for (int i = 0;i < local;i++)  //遍历
			cur = cur->next;
		return cur;
	}
	else
		return NULL;
}

//逆转表(方法1，牺牲时间换取空间，速度很慢，时间复杂度为0(n^2))
void Reverse1()
{
	STUDENT temp;
	for (int i =0;i<g_iCount/2;i++)
	{
		LPLINKLIST node1 = GetNode(i);                //获取节点
		LPLINKLIST node2 = GetNode(g_iCount - i - 1);
		temp = node1->s;                              //交换数据项，千万别交换节点
		node1->s = node2->s;
		node2->s = temp;
	}
}

//逆转表(方法2，牺牲空间换取时间，速度很快，时间复杂度为0(n))
void Reverse2()
{
	LINKLIST tHead = { 0 };     //临时头节点
	LPLINKLIST cur = head.next;
	for (int i = 0;i < g_iCount;i++)
	{
		LPLINKLIST l = (LPLINKLIST)malloc(sizeof(_link_list));
		memcpy(l, cur, sizeof(_link_list));   //这里一定要复制，而不是赋值=，如果是=，则两指针指向的是同一个节点
		l->next = tHead.next;        //这里采用头添加，结束后，就会自动逆转过来了         
		tHead.next = l;
		cur = cur->next;
	}
	cur = head.next;
	LPLINKLIST temp = tHead.next;
	for (int i = 0;i < g_iCount;i++) 
	{
		memcpy(cur, temp, sizeof(_link_list));   //因为头添加已经是逆转了，所以直接复制到原来的链表中即可
		temp = temp->next;
		cur = cur->next;
	}
}

//修改数据
void Rewrite(int local, STUDENT s)
{
	if (local < g_iCount && local >= 0)   //当要删除的位置合理的情况下才能删除
	{
		LPLINKLIST cur = head.next;
		for (int i = 0;i < local;i++)    //遍历到指定位置
			cur = cur->next;
		cur->s = s;                       //修改
	}
}

//清空链表
void ClearLinkList()
{
	LPLINKLIST cur = head.next;
	for (int i = 0;i<g_iCount;i++)
	{
		LPLINKLIST temp;
		temp = cur;         //保存前一个节点的地址 
		cur = cur->next;    //向前查找下一个节点
		free(temp);         //释放前一个节点的空间
	} 
	head.next = NULL;       //最后让头节点指向NULL
}

//遍历数据
void Print()
{
	LPLINKLIST cur = &head;
	while (cur = cur->next)
		printf("%d - %s\n", cur->s.id, cur->s.name);
}

File: JAVA/LinkList.java

/*
* author: acme
*   date: 2017-1-18
*  blogs: http://blog.csdn.net/qq_18297675
* update: 2017-1-25
*   info: add tail ptr,improve the efficiency of adding data,change node class become inner private class
*/

public class LinkList<T> {
	
	private class Node implements Cloneable{  //实现该接口后，可进行对象的复制
		
		public T data;
		public Node next;
		Node(T d) {
			data = d;
		} 
		
		//重写clone  用来进行浅复制
		public Object clone() {
			Node node = null;
			try {
				node = (Node)super.clone();
			} catch (CloneNotSupportedException e) {
				e.printStackTrace();
			}
			return node;
		}
	}
	
	private Node head;    //头节点
	private Node tail;    //尾指针
	private int iCount;   //链表大小
	
	public LinkList() {
		initList();
	}
	
	private void initList() {
		head = new Node(null);    		//头节点
		tail = new Node(null);    		//尾指针    
		iCount = 0;
	}
	
	//判断表是否为空
	public boolean isEmpty() {
		return iCount == 0 ? true : false;
	}
	
	//获取元素个数
	public int size() {
		return iCount;
	}
	
	//添加数据（头添加）
	public void addFromHead(T e) {		
		Node node = new Node(e);   //用e来创建新的节点
		if (isEmpty())//如果链表为空，第一次添加的时候，就让尾指针指向第一个添加的节点
			tail.next = node;
		node.next = head.next;
		head.next = node;
		++iCount;
	}
	
	//添加数据（尾添加）
	public void addFromTail(T e) {		
		if (isEmpty())
			addFromHead(e);
		else {
			Node node = new Node(e);
			tail.next.next = node;//让最后一个节点指向新添进来的节点
			tail.next = node;	  //修改tail的指向
			++iCount;
		}
	}
	
	//插入数据（指定位置）
	public void insert(int local, T e) {	
		if (local < iCount && local >= 0) {
			Node cur = head;   //注意，这里千万不能是m_head->next,因为如果当前链表为空插入的位置
								  //是0，则会直接执行cur->next这个表达式，解引用空指针会出错
			Node node = new Node(e);
			for (int i = 0;i < local;i++, cur = cur.next); //遍历到local这个位置
			node.next = cur.next;
			cur.next = node;       //这里的操作就和头添加一样,就是改变指向
			++iCount;
		}
		else if (local == iCount)//如果是插入尾部，直接调用AddFromTail函数即可
			addFromTail(e);
	}
	
	//删除数据 （指定位置）
	public void deleteFromLocal(int local) {		
		if (local < iCount && local >= 0) {//当删除的位置合法的时候才进行删除操作
			Node pre = head;
			Node cur = head.next;
			for (int i = 0;i < local;i++, cur = cur.next, pre = pre.next);  //遍历到指定位置
			pre.next = cur.next;  //例如:pre->cur->next,现在直接让pre->next
			if(pre.next == null)  //如果是删除最后一个元素，则需要修改尾指针
				tail.next = pre;
			--iCount;
		}
	}
	
	//删除数据 （指定元素）
	public void deleteFromElement(T e) {			
		int local = getLocal(e);
		deleteFromLocal(local);
	}
	
	//根据指定元素查找，返回位置
	public int searchFromElmemt(T e) {				
		return getLocal(e);
	}
	
	//根据位置查找指定元素  
	public T searchFromLocal(int local) throws Exception {		
		Node node = getNode(local);
		if (node != null) {
			return node.data;
		}
		else
			throw new Exception("查找的位置不存在，查找失败.");
	}
	
	////逆转表(方法1，牺牲时间换取空间，速度很慢，时间复杂度为0(n^2))
	public void reverse1() {						
		T temp;
		for (int i = 0;i < iCount / 2;i++) {
			Node node1 = getNode(i);                   //获取节点
			Node node2 = getNode(iCount - i - 1);
			temp = node1.data;                         //交换数据项，千万别交换节点
			node1.data = node2.data;
			node2.data = temp; 
		}
	}
	
	//逆转表(方法2，牺牲空间换取时间，速度很快，时间复杂度为0(n))
	public void reverse2() {
		Node tHead = new Node(null);     //临时头节点
		Node cur = head.next;
		for (int i = 0;i < iCount;i++) {
			Node node = new Node(cur.data);   //创建新的节点
			node.next = tHead.next;			  //这里采用头添加，结束后，就会自动逆转过来了         
			tHead.next = node;
			cur = cur.next;
		}
		head = (Node)tHead.clone();   //因为头添加已经是逆转了，所以直接复制到原来的链表中即可
	}
	
	//遍历元素
	public void print() {
		Node cur = head.next;
		for (int i = 0; i < iCount; i++,cur = cur.next) {
			System.out.println(cur.data);
		}
	}
	
	//复制表
	public LinkList<T> copyList() {	//不能直接用clone方法，必须要一个个节点复制
		LinkList<T> list = new LinkList<>();
		Node cur = head.next;
		for (int i = 0; i < iCount; i++,cur = cur.next) {
			list.addFromTail(cur.data);  //加了尾指针，头添加和尾添加都一样了
		}
		return list;
	}
	
	//修改元素
	public void rewrite(int local, T e) {			
		Node node = getNode(local);
		if (node != null)
			node.data = e;
	}
	
	//清空链表
	public void clearLinkList() {                   
		head.next = null;       //因为java有垃圾回收机制，所以不需要手动释放节点，只要让头节点为空就好
		tail.next = null;
		iCount = 0;
	}
	
	//获取指定位置的节点
	public Node getNode(int local) {            
		if (local < iCount && local >= 0) {//当获取的位置合法的时候才进行获取操作
			Node cur = head.next;
			for (int i = 0;i < local;i++, cur = cur.next);
			return cur;
		}
		else
			return null;
	}
	
	//返回元素位置
	private int getLocal(T e) { 						
		Node cur = head.next;
		for (int i = 0;i < iCount;i++, cur = cur.next)
			if (cur.data == e)
				return i;
		return -1;
	}
	
	//升序
	public void sortByAsc() {
		selectSort(true);
	}
	
	//降序
	public void sortByDesc() {
		selectSort(false);
	}
	
	//选择排序法
	private void selectSort(boolean bAsc) {
		for (int i = 0;i < iCount;i++) {
			Node node1 = getNode(i);
			for (int j = i + 1;j<iCount;j++) {
				Node node2 = getNode(j);
				T temp;
				if (bAsc) {//升序
					if ((int)node1.data > (int)node2.data) {  //排序只对整型有效
						temp = node1.data;
						node1.data = node2.data;
						node2.data = temp;
					}
				}
				else  {//降序
					if ((int)node1.data < (int)node2.data) {
						temp = node1.data;
						node1.data = node2.data;
						node2.data = temp;
					}
				}
			}
		}
	}
	
}
